aws_lambda_builders/workflows/nodejs_npm/utils.py

```diff
@@ -48,3 +64,124 @@ def abspath(self, path):
 
     def is_windows(self):
         return platform.system().lower() == "windows"
+
+
+class DependencyUtils(object):
+
+    """
+    Collection of helper functions for managing local NPM dependencies
+    """
+
+    @staticmethod
+    def get_local_dependencies(manifest_path, osutils):
+        """
+        Helper function to extract all local dependencies in a package.json manifest
+        """
+
+        with osutils.open_file(manifest_path) as manifest_file:
+            manifest = json.loads(manifest_file.read())
+
+            if "dependencies" in manifest:
+                return dict(
+                    (k, v) for (k, v) in manifest["dependencies"].items() if DependencyUtils.is_local_dependency(v)
+                )
+            else:
+                return {}
+
+    @staticmethod
+    def is_local_dependency(path):
+        """
+        Helper function to check if package dependency is a local package
+        """
+
+        try:
+            return path.startswith("file:") or path.startswith(".")
+        except AttributeError:
+            return False
+
+    @staticmethod
+    def package_local_dependency(
```

Comment:
We may need a bit of refactoring and cleaning of this function (the TODOs).
Between this one and `execute`, we are calling the `npm pack` action in three different places.

Reply:
Two of the `npm pack` calls were combined into a common function (where we're packing a module and then extracting the archive to a scratch directory -- "shipping"). The 3rd instance of `npm pack` is to repackage a local dependency after having "shipped" it to the scratch directory and recursively searched it for its own local deps. This final pack doesn't extract the local dep, so it's slightly different from the previous 2 calls.

Reply:
I haven't addressed the `TODOs` here yet.

Reply:
All right, I'm checking to see if we can do the whole packaging in one pass.

Reply:
Here's a tentative solution, what do you think?
It should also handle already processed and circular dependencies

https://gist.github.com/mgrandis/2bc805f4bf021905fb1cd71212fab144

Reply:
In the current situation, we are uploading the modified `package.json` files.
We may want to upload the original ones: make a `.bak` and revert to it in the final `NodejsNpmrcCleanUpAction`.

Reply:
Looks great and it's much cleaner! Thank you very much. I knew there had to be a way to simplify it, but I couldn't figure it out.

I think reverting to the original `package.json` might also fix #187. Do I need to reproduce your gist, or can you push those updates here?

Reply:
As you wish, you will have to give me access to your repository if you want me to push.
I'm also fine with you copy/pasting the code on your side, no worries.

Good catch for the #187 issue, I think it would boil down to:
* Make a copy of the `package.json` to `package.json.bak` at L30 of the `utils.py` gist (it will be packed by npm);
* Update the `NodejsNpmrcCleanUpAction::execute` to search for `package.json.bak` files in the final artifacts dir/subdirs, copy each to `package.json` and remove the `.bak` file;
* Rename `NodejsNpmrcCleanUpAction` to `NodejsNpmCleanUpAction` (remove the `rc` part).

We will need tests for all this ;)

Reply:
@cBiscuitSurprise, have you had time working on this?

Reply:
@mgrandis I have not. Been pretty busy / distracted. I plan to, unless you have some time.

Reply:
No problem, I was just checking in to see if you needed any help.

Reply:
Sent you a PR to update all this: https://github.com/cBiscuitSurprise/aws-lambda-builders/pull/1